File: src/libs/cpp/functions/smoothstep.cpp
#include "smoothstep.hpp"

void smoothstep(Context &context) noexcept
{
	// Smoothstep interpolation between two values.
	auto params = std::get<std::vector<Value>>(context.stack.pop());
	double x = params[0].to_number();
	double min = params[1].to_number();
	double max = params[2].to_number();

	const double range = max - min;
	double value = (std::min(std::max(min, x), max) - min) / range;
	value = value * value * (3 - 2 * value);

	context.stack.push(value * range + min);
}

File: src/runtime/functions/smoothstep.lua
return function(vm)
	local v = vm.pop()
	local value, min, max = std.num(v[1]), std.num(v[2]), std.num(v[3])

	local range = max - min
	value = (math.min(math.max(min, value), max) - min) / range
	value = value * value * (3.0 - 2.0 * value)
	vm.push(value * range + min)
end
